### PR TITLE
Add 'batch' subcommand to sawadm

### PR DESCRIPTION
**Pull request by isabeltomb**
Add the `batch` subcommand to sawadm. The `batch` subcommand has a `list` subcommand which lists all existing batches in the current blockchain. `batch list` has optional flags for specifying the authorizer of the request, the output format of the batches, and the URL of the Validator's REST API.

### CODE REVIEW

File: adm/Cargo.toml [hunk 1]
```diff
@@ -29,9 +29,11 @@ clap = ">=2.29.0"
 libc = ">=0.2.35"
 lmdb-zero = ">=0.4.1"
 protobuf = "2.0"
+sawtooth = { version = "0.6", features = ["client-rest"] }
```
Comment (peterschwarz) [inline adm/Cargo.toml:32]: This should be an optional dependency, for now, and it should be included under an experimental feature below.  Here, it can just say optional, where as when it's use, the experimental feature (possibly just called `client-cli`) can use it as such:

```
client-cli = ["sawtooth/client-rest", ...]
```

File: adm/src/main.rs [hunk 2]
```diff
@@ -20,6 +20,7 @@ extern crate clap;
 extern crate libc;
 extern crate lmdb_zero;
 extern crate protobuf;
+extern crate sawtooth;
```
Comment (peterschwarz) [inline adm/src/main.rs:23]: since this is now 2018 edition, these can be omitted.

File: adm/src/commands/mod.rs [hunk 3]
```diff
@@ -15,6 +15,7 @@
  * ------------------------------------------------------------------------------
  */
 
+pub mod batch;
```
Comment (peterschwarz) [inline adm/src/commands/mod.rs:18]: This should be behind the experimental feature defined in the Cargo toml, as should the uses and the clap args in main.rs.

**Review by peterschwarz — CHANGES_REQUESTED** [review]

File: adm/Cargo.toml (hunk 1 above)
Comment (isabeltomb) [reply adm/Cargo.toml]: Fixed

File: adm/src/main.rs (hunk 2 above)
Comment (isabeltomb) [reply adm/src/main.rs]: Fixed

File: adm/src/commands/mod.rs (hunk 3 above)
Comment (isabeltomb) [reply adm/src/commands/mod.rs:19]: Fixed

File: adm/src/main.rs [hunk 4]
```diff
@@ -17,14 +17,8 @@
 
 #[macro_use]
 extern crate clap;
-extern crate libc;
```
Comment (peterschwarz) [inline adm/src/main.rs:20]: Could you move these removals to another commit?

File: adm/src/main.rs [hunk 5]
```diff
@@ -113,6 +109,17 @@ fn parse_args<'a>() -> ArgMatches<'a> {
             (@arg ignore_required_settings: --("ignore-required-settings")
              "skip the check for settings that are required at genesis (necessary if using a
               settings transaction family other than sawtooth_settings)"))
+        (@subcommand batch =>
```
Comment (peterschwarz) [inline adm/src/main.rs:112]: This should also be behind the feature

File: adm/src/commands/batch.rs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+            } else {
+                col_string += &" ".repeat(*len);
+            }
+        }
+        println!("{}", col_string);
+    }
+}
+
+#[derive(Debug, Serialize)]
+struct JsonYamlBatch {
```
Comment (peterschwarz) [inline adm/src/commands/batch.rs:164]: Maybe this could be called StructuredOutputBatch

File: adm/src/main.rs (hunk 5 above)
Comment (isabeltomb) [reply adm/src/main.rs]: Fixed

File: adm/src/commands/batch.rs (hunk 6 above)
Comment (isabeltomb) [reply adm/src/commands/batch.rs]: Fixed

File: adm/src/main.rs (hunk 4 above)
Comment (isabeltomb) [reply adm/src/main.rs:20]: Moved to separate commit

File: adm/src/main.rs [hunk 7]
```diff
@@ -107,5 +112,47 @@ fn parse_args<'a>() -> ArgMatches<'a> {
...
+                                "specify the user to authorize\n\
+                                request",
+                            ),
+                    )
+                    .arg(
+                        Arg::with_name("format")
+                            .long("format")
+                            .short("F")
+                            .takes_value(true)
+                            .value_name("csv,json,yaml,default")
```
Comment (ltseeley) [inline adm/src/main.rs:152]: This should be `.possible_values(...)` https://docs.rs/clap/2.33.3/clap/struct.Arg.html#method.possible_values

File: adm/src/main.rs [hunk 8]
```diff
@@ -107,5 +112,47 @@ fn parse_args<'a>() -> ArgMatches<'a> {
...
+                        "identify the URL of the validator's\n\
+                            REST API (default:\n\
+                            http://localhost:8008)",
+                    ))
+                    .arg(
+                        Arg::with_name("username")
+                            .long("user")
+                            .short("u")
+                            .takes_value(true)
+                            .value_name("USERNAME[:PASSWORD]")
```
Comment (ltseeley) [inline adm/src/main.rs:141]: We should probably just add the format to the help text instead of with `value_name`.

File: adm/src/commands/batch.rs [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+                CliError::EnvironmentError(format!("Failed to create client: {}", err))
+            })?;
+    } else {
+        client = RestApiSawtoothClientBuilder::new()
+            .with_url(&url)
+            .build()
+            .map_err(|err| {
+                CliError::EnvironmentError(format!("Failed to create client: {}", err))
+            })?;
+    }
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply adm/src/commands/batch.rs]: fixed

File: adm/src/commands/batch.rs [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+            )
+        }
+    } else {
+        let data = parse_batches(batches)?;
+        print_table(data);
+    }
+    Ok(())
+}
+
+fn parse_batches(
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply adm/src/commands/batch.rs]: Changed function name and added doc comment

File: adm/src/commands/batch.rs [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+        for batch in batches {
+            println!(
+                "{}",
+                serde_json::to_string_pretty(&StructuredOutputBatch::from(&batch.unwrap()))
+                    .map_err(|err| CliError::EnvironmentError(format!(
+                        "Cannot format batches into json: {}",
+                        err
+                    )))?
+            )
+        }
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply adm/src/commands/batch.rs]: Fixed

File: adm/src/commands/batch.rs [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+    if format == "csv" {
+        let data = parse_batches(batches)?;
+        for row in data {
+            println!("{}", row.join(","))
+        }
+    } else if format == "json" {
+        for batch in batches {
+            println!(
+                "{}",
+                serde_json::to_string_pretty(&StructuredOutputBatch::from(&batch.unwrap()))
```
[the remark this answers is not in the document]
Comment (isabeltomb) [reply adm/src/commands/batch.rs]: Fixed

File: adm/src/main.rs (hunk 7 above)
Comment (isabeltomb) [reply adm/src/main.rs]: Fixed

File: adm/src/main.rs (hunk 8 above)
Comment (isabeltomb) [reply adm/src/main.rs]: Fixed

**Review by ltseeley — APPROVED** [review]

**Review by peterschwarz — APPROVED** [review]
